### PR TITLE
jit(arm64): support for memory loads/stores

### PR DESCRIPTION
**Pull request by mathetake**
part of #187

### CODE REVIEW

**Comment by mathetake** [comment]
ok done with store

File: wasm/jit/jit_arm64.go [hunk 1]
```diff
@@ -1773,36 +1780,211 @@ func (c *arm64Compiler) compileGe(o *wazeroir.OperationGe) error {
 	return nil
 }
 
+// compileLoad implements compiler.compileLoad for the amd64 architecture.
```
Comment (nullpo-head) [inline wasm/jit/jit_arm64.go:1783]:
```suggestion
// compileLoad implements compiler.compileLoad for the arm64 architecture.
```

File: wasm/jit/jit_arm64.go [hunk 2]
```diff
@@ -1773,36 +1780,211 @@ func (c *arm64Compiler) compileGe(o *wazeroir.OperationGe) error {
...
+		targetSizeInBytes = 32 / 8
+	case wazeroir.UnsignedTypeF64:
+		loadInst = arm64.AFMOVD
+		isFloat = true
+		targetSizeInBytes = 64 / 8
+	}
+	return c.compileLoadImpl(o.Arg.Offset, loadInst, targetSizeInBytes, isFloat)
 }
 
+// compileLoad8 implements compiler.compileLoad8 for the amd64 architecture.
```
Comment (nullpo-head) [inline wasm/jit/jit_arm64.go:1810]:
```suggestion
// compileLoad8 implements compiler.compileLoad8 for the arm64 architecture.
```

File: wasm/jit/jit_arm64.go [hunk 3]
```diff
@@ -1773,36 +1780,211 @@ func (c *arm64Compiler) compileGe(o *wazeroir.OperationGe) error {
...
+		// TODO: looks like the assembler cannot emit 32-bit variant of LDRSB.
+		// Differentiate 32-bit vs 64-bit after #233.
+		loadInst = arm64.AMOVB
+	case wazeroir.SignedUint32, wazeroir.SignedUint64:
+		loadInst = arm64.AMOVBU
+	}
+	return c.compileLoadImpl(o.Arg.Offset, loadInst, 1, false)
 }
 
+// compileLoad16 implements compiler.compileLoad16 for the amd64 architecture.
```
Comment (nullpo-head) [inline wasm/jit/jit_arm64.go:1824]:
```suggestion
// compileLoad16 implements compiler.compileLoad16 for the arm64 architecture.
```

File: wasm/jit/jit_arm64.go [hunk 4]
```diff
@@ -1773,36 +1780,211 @@ func (c *arm64Compiler) compileGe(o *wazeroir.OperationGe) error {
...
+		// TODO: looks like the assembler cannot emit 32-bit variant of LDRSH.
+		// Differentiate 32-bit vs 64-bit after #233.
+		loadInst = arm64.AMOVH
+	case wazeroir.SignedUint32, wazeroir.SignedUint64:
+		loadInst = arm64.AMOVHU
+	}
+	return c.compileLoadImpl(o.Arg.Offset, loadInst, 16/8, false)
 }
 
+// compileLoad32 implements compiler.compileLoad32 for the amd64 architecture.
```
Comment (nullpo-head) [inline wasm/jit/jit_arm64.go:1838]:
```suggestion
// compileLoad32 implements compiler.compileLoad32 for the arm64 architecture.
```

File: wasm/jit/jit_arm64.go [hunk 5]
```diff
@@ -1773,36 +1780,211 @@ func (c *arm64Compiler) compileGe(o *wazeroir.OperationGe) error {
...
+		loadInst,
+		reservedRegisterForMemory, offsetReg,
+		resultRegister,
+	)
+
+	c.locationStack.pushValueLocationOnRegister(resultRegister)
+	return nil
+}
+
+// compileStore implements compiler.compileStore for the amd64 architecture.
```
Comment (nullpo-head) [inline wasm/jit/jit_arm64.go:1876]:
```suggestion
// compileStore implements compiler.compileStore for the arm64 architecture.
```

File: wasm/internal/wazeroir/operations.go [hunk 6]
```diff
@@ -552,6 +552,8 @@ func (o *OperationStore) Kind() OperationKind {
 }
 
 type OperationStore8 struct {
+	// TODO: Semantically Type doesn't affect operation,
```
Comment (codefromthecrypt) [inline wasm/internal/wazeroir/operations.go:555]: supernit: TODO rarely deserve 2 lines

File: wasm/jit/engine_test.go [hunk 7]
```diff
@@ -71,6 +71,8 @@ func TestVerifyOffsetValue(t *testing.T) {
 	// Offsets for wasm.MemoryInstance
 	var memoryInstance wasm.MemoryInstance
 	require.Equal(t, int(unsafe.Offsetof(memoryInstance.Buffer)), memoryInstanceBufferOffset)
+	// "+8" because the slice header is laid out as {Data uintptr, Len, int64, Cap int64} on memory.
```
Comment (codefromthecrypt) [inline wasm/jit/engine_test.go:74]: thanks!

File: wasm/jit/jit_arm64.go (hunk 2 above)
Comment (codefromthecrypt) [inline wasm/jit/jit_arm64.go:1810]: s/amd64/arm64/g in this file

File: wasm/jit/jit_arm64.go [hunk 8]
```diff
@@ -1773,36 +1780,211 @@ func (c *arm64Compiler) compileGe(o *wazeroir.OperationGe) error {
...
 }
 
+// compileLoad16 implements compiler.compileLoad16 for the amd64 architecture.
 func (c *arm64Compiler) compileLoad16(o *wazeroir.OperationLoad16) error {
-	return fmt.Errorf("TODO: unsupported on arm64")
+	var loadInst obj.As
+	switch o.Type {
+	case wazeroir.SignedInt32, wazeroir.SignedInt64:
+		// TODO: looks like the assembler cannot emit 32-bit variant of LDRSH.
+		// Differentiate 32-bit vs 64-bit after #233.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_arm64.go:1830]: good note

File: wasm/jit/jit_arm64.go [hunk 9]
```diff
@@ -1773,36 +1780,211 @@ func (c *arm64Compiler) compileGe(o *wazeroir.OperationGe) error {
...
 }
 
+// compileStore32 implements compiler.compileStore32 for the amd64 architecture.
 func (c *arm64Compiler) compileStore32(o *wazeroir.OperationStore32) error {
-	return fmt.Errorf("TODO: unsupported on arm64")
+	return c.compileStoreImpl(o.Arg.Offset, arm64.AMOVW, 32/8)
+}
+
+// compileStoreImpl implements compleStore* variants for arm64 architecture.
+func (c *arm64Compiler) compileStoreImpl(offsetArg uint32, storeInst obj.As, targetSizeInBytes int64) error {
```
Comment (codefromthecrypt) [inline wasm/jit/jit_arm64.go:1913]: is int64 excessive
```suggestion
func (c *arm64Compiler) compileStoreImpl(offsetArg uint32, storeInst obj.As, targetSizeInBytes uint32) error {
```

File: wasm/jit/jit_arm64.go [hunk 10]
```diff
@@ -1773,36 +1780,211 @@ func (c *arm64Compiler) compileGe(o *wazeroir.OperationGe) error {
...
+	return nil
+}
+
+// compileMemoryAccessOffsetSetup pops the top value from the stack (called "base"), stores "base + offsetArg + targetSizeInBytes"
+// into a register, and returns the stored register. We call the result "offset" because we access the memory
+// as memory.Buffer[offset: offset+targetSizeInBytes].
+//
+// Note: this also emits the instructions to check the out of bounds memory access.
+// In other words, if the offset+targetSizeInBytes exceeds the memory size, the code exits with jitCallStatusCodeMemoryOutOfBounds status.
+func (c *arm64Compiler) compileMemoryAccessOffsetSetup(offsetArg uint32, targetSizeInBytes int64) (offsetRegister int16, err error) {
```
Comment (codefromthecrypt) [inline wasm/jit/jit_arm64.go:1943]: uint32 or less. ack when adding you may want to cast, but it is distracting here as I don't think the value will be more than single or double digits right?

```suggestion
func (c *arm64Compiler) compileMemoryAccessOffsetSetup(offsetArg uint32, targetSizeInBytes uint32) (offsetRegister int16, err error) {
```

File: wasm/jit/jit_arm64.go [hunk 11]
```diff
@@ -2088,66 +2270,119 @@ func (c *arm64Compiler) compileInitializeReservedStackBasePointerRegister() erro
...
+			reservedRegisterForMemory,
+		)
+	}
+}
+
 // compileModuleContextInitialization adds instructions to initialize engine.ModuleContext's fields based on
 // engine.ModuleContext.ModuleInstanceAddress.
 // This is called in two cases: in function preamble, and on the return from (non-Go) function calls.
 func (c *arm64Compiler) compileModuleContextInitialization() error {
 	// Obtain the free registers to be used in the followings.
```
Comment (codefromthecrypt) [inline wasm/jit/jit_arm64.go:2288]: this comment says nothing, so you can delete it

**Review by codefromthecrypt — APPROVED** [review]
Good job!

**Review by nullpo-head — APPROVED** [review]
awesome!

File: wasm/jit/jit_arm64.go (hunk 9 above)
Comment (mathetake) [reply wasm/jit/jit_arm64.go:1911]: agreed, but in anyway we end up casting this to int64 to pass to assembler, so I intentionally use int64 here. I will leave as-is! Thanks!

**Pull request merged**